numpy/testing/utils.py

```diff
@@ -1937,3 +1964,150 @@ def __exit__(self, *exc_info):
                 mod.__warningregistry__.clear()
             if mod in self._warnreg_copies:
                 mod.__warningregistry__.update(self._warnreg_copies[mod])
+
+
+class suppressed_warnings(object):
+    """
+    Context manager doing much the same as ``warnings.catch_warnings``.
+    However, it also provides a suppression mechanism, meaning that instead
+    of setting a warning to "ignore", it is possible to add a suppression.
```

Comment:
I have updated the text a bit pointing to the python issue, I hope it is somewhat better now. There were also some smaller issues in how the filters worked.

I do not claim this thing is all that great, but I *really* like the end result of this branch, I could just rebase that other new FutureWarning feature on here, and actually find all the tests starting to be flaky immidiatly.